### PR TITLE
[DSCP-393] replace config type with vinyl records

### PR DESCRIPTION
**Pull request by pasqu4le**
### Description

This PR aims at replacing as many configuration-specific data types as possible with equivalent `loot-config` types (`vinyl` records).

I have tried to make as little changes as possible to usage logic, CLI args handling and configuration hierarchy; the differences there are only the ones that seemed necessary.

### YT issue

https://issues.serokell.io/issue/DSCP-393

### Checklist

Hint: a perfect PR has all the checkmarks set.

Possible related changes (conditional):
- [x] I added tests if required, in case they are:
  - Covering new introduced functionality (feature).
  - Regression tests preventing the bug from silently reappearing again (bugfix).
- [x] I have checked the [sample config](../tree/master/docs/config-full-sample.yaml) and [launch scripts](../tree/master/scripts/launch) and updated them if it was necessary (especially if executables or CLIs were changed).
- [x] I have checked READMEs and changed them accordingly if it's necessary (the main [README.md](../tree/master/README.md) as well as subproject READMEs for all related executables).
- [x] If a public API structure or/and data serialization was changed, I made sure that these changes are reflected in the:
  - [Swagger](../tree/master/specs/disciplina) API specs.
  - Any [related documentation](../tree/master/docs/api-types.md).
- [x] If any public documentation was written, I have spellchecked it.

Stylistic (obligatory):
- [x] My commit history is clean, descriptive and do not contain merge or revert commits or commits like "WIP".
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).
- [x] My code is documented (haddock) and these docs are decent (full enough and spellchecked).

### CODE REVIEW

**Review by pasqu4le — COMMENTED** [review]
I think I left a comment for every change (or lack thereof) that I had more doubt with.
Hope it can be of any help.

**Comment by pasqu4le** [comment]
@flyingleafe I just made a small fixup commit while revisiting the whole configuration.

Apart from the 2 doubts I expressed already in the comments, I think that with a couple more changes this PR could close the parent issue, [DSCP-387](https://issues.serokell.io/issue/DSCP-387), as well.

Specifically:
- `EducatorBotParams` should probably be modified into a `tree`, much like `AppDirParam`, because right now when it is not `enabled` we still have to define, unnecessarily, `seed` and `operationsDelay` (hence the need of the silly default: "Memes generator")
- `MultiEducator` has two functions that define default values (`defSqliteParams` and `defBotParams`) that are a copy-paste from `Educator`, we could simply make it export them
- something similar happens for `Educator`'s `defBaseKeyParams` and `Witness`' `keysParams` with the only difference that the latter contains the former

File: witness/src/Dscp/Witness/CLI.hs [hunk 1]
```diff
@@ -85,13 +91,20 @@ metricsServerParser =
...
 ---------------------------------------------------------------------------
 
-committeeParamsParser :: Parser CommitteeParams
-committeeParamsParser =
-    combine <$> nParser <*> optional commSecretParser
+committeeKeyParamsParser :: OptModParser CommitteeKeyParams
+committeeKeyParamsParser = #params .:+
+    (#participantN .:: nParser <*<
+     (#paramsType .:: ("open" <$ openP)) <|>
+     ((#paramsType .:: pure "closed" <*<
```
Comment (flyingleafe) [inline witness/src/Dscp/Witness/CLI.hs]: This pattern seems to be repeated in CLI parsers for sum types, and it's not particularly pretty. Can we make a helper which does the same thing?

Comment (flyingleafe) [reply witness/src/Dscp/Witness/CLI.hs]: Also, why not just do `#paramsType .:: ("open" <$ openP <|> pure "closed")` and parse `#closed` branch in each case? `OptModParser`s never fail anyway, and in case of `open` committee redundant branch will be discarded during config finalization. Basically, you do exactly this in `witnessKeyParamsParser`

File: witness/src/Dscp/Witness/CLI.hs [hunk 2]
```diff
@@ -101,17 +114,26 @@ committeeParamsParser =
...
-    modifying (Committee <$> committeeParamsParser) <|>
-    over _Basic <$> baseKeyParamsParser "witness"
+witnessKeysParamsParser :: OptModParser WitnessKeysParams
+witnessKeysParamsParser = #params .:+
+    (#paramsType %:: (basicP <|> committeeP) <*<
+     #basic      .:- baseKeyParamsParser "witness" <*<
+     #committee  .:- committeeKeyParamsParser
+    )
+  where
+    basicP = flag' (const "basic") $
```
Comment (flyingleafe) [inline witness/src/Dscp/Witness/CLI.hs]: Why do you return `const` functions, if you can just return values and use `.::` instead of `%::`?

File: witness/src/Dscp/Witness/Config.hs [hunk 3]
```diff
@@ -18,26 +18,25 @@ module Dscp.Witness.Config
...
     '[ "witness" ::<
-       '[ "logging" ::: LoggingParams
-        , "db" ::: RocksDBParams
-        , "network" ::: NetServParams
-        , "keys" ::: WitnessKeyParams
+       '[ "logging" ::< LoggingParams
+        , "db" ::< RocksDBParams
+        , "network" ::< NetServParams
+        , "keys" ::< WitnessKeysParams
         , "api" ::: Maybe ServerParams
```
[the remark this answers is not in the document]
Comment (flyingleafe) [reply witness/src/Dscp/Witness/Config.hs]: Well, `Maybe` is just a sum type, right? For such cases, we can come up with some `ConfigMaybe cfg`, which is just `Maybe`, but via vinyl configs. This way we'll also allow the situation when the API params are defined in `config1`, but then API is disabled in `config2`, and then re-enabled again via CLI param.

**Review by flyingleafe — COMMENTED** [review]
Great job! I only have a bunch of not-so-major comments and one question.
Also, thanks for your inline comments, they really made the review easier.

File: witness/src/Dscp/Witness/CLI.hs (hunk 1 above)
Comment (pasqu4le) [reply witness/src/Dscp/Witness/CLI.hs]: we cannot do that, because even tho in practice it could work, this would make it so that, in this case, `"closed"` is selected even when we do not specify **any** CLI arguments.

This does in fact break the proposition that we test for, in this case:
>spec_WitnessCliParamsNoDefaults :: Spec
>spec_WitnessCliParamsNoDefaults = describe "Witness node CLI interface" $
>    it "should not yield any default values if no CLI params are provided" $
>        (runCliArgs witnessConfigParser [] <*> pure mempty) \`shouldBe\` Just mempty

does not hold anymore.

In the case of `witnessKeyParamsParser` there is a difference: both the selection possibilities have their own CLI argument, and I think it is necessary here.
Basically for `committeeKeyParamsParser` I was trying to avoid to add another CLI argument just to make the "closed" selection and let the `--comm-sec` option handle that as well, so one can define the content of the `closed` branch and select it at the same time. 

Agreed on the ugliness and the helper idea, I'll try to make one.

File: witness/src/Dscp/Witness/CLI.hs (hunk 2 above)
Comment (pasqu4le) [reply witness/src/Dscp/Witness/CLI.hs]: I honestly don't recall my reasoning behind this, I think I was trying something else and left it like this by mistake.
You are right, I'll fix it.

File: educator/src/Dscp/Educator/Web/Bot/Params.hs [hunk 4]
```diff
@@ -1,30 +1,24 @@
...
-import Dscp.Util.Aeson ()
-
 -- | Which params to use when launching bot.
 -- Bool flag is used instead of `Maybe` value or custom sum type
 -- to allow for specifying default values for bot params in
 -- default config even though bot should be disabled by default.
-data EducatorBotParams = EducatorBotParams
-    { ebpEnabled         :: !Bool
+type EducatorBotParams =
+   '[ "enabled" ::: Bool
```
[the remark this answers is not in the document]
Comment (pasqu4le) [reply educator/src/Dscp/Educator/Web/Bot/Params.hs]: I suggested this as well in my latest comment (after the "review").

I also put there two more suggestion about merging identical functions, could you take a look at those as well?

File: witness/src/Dscp/Witness/Config.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (pasqu4le) [reply witness/src/Dscp/Witness/Config.hs]: Right! How did I not think of that?

Sure thing, I'll write `ConfigMaybe` and convert `ServerParams`.

File: educator/src/Dscp/DB/SQLite/Types.hs [hunk 5]
```diff
@@ -1,56 +1,61 @@
...
+type SQLiteDBModeRec = Config SQLiteDBMode
+type SQLiteDBModeRecP = PartialConfig SQLiteDBMode
 
-makePrisms ''SQLiteDBMode
 
-data SQLiteParams = SQLiteParams
-    { sdpMode :: SQLiteDBMode
-    } deriving (Show, Eq)
+-- | Wrapper for 'SQLiteDBMode', this helps passing the vinyl record around
+type SQLiteParams =
```
[the remark this answers is not in the document]
Comment (pasqu4le) [reply educator/src/Dscp/DB/SQLite/Types.hs]: One more thing I was just thinking: in most cases there is not need for both the "wrapper" and the "content only", so it may be a good idea to just define the second one inside the first.
At least we would avoid making unnecessary similarly-named types.

File: educator/src/Dscp/Educator/Web/Bot/Params.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (flyingleafe) [reply educator/src/Dscp/Educator/Web/Bot/Params.hs]: Wow, somehow I missed that last comment of yours, how did it happen :man_facepalming: 
Yes, I like your proposed additional changes. Let's do them.

**Comment by pasqu4le** [comment]
@flyingleafe everything should be done now, also I noticed later that a reply I wrote was not visible in the PR discussion page, but only in the "Files changes" page, for some reason.

Anyway, if you have not read it, it [explains the weird sum-types wrappers](https://github.com/DisciplinaOU/disciplina/pull/346/files#r243335266).
As promised, a shorter explanation is on in a comment as well.

File: educator/src/Dscp/DB/SQLite/Types.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (flyingleafe) [reply educator/src/Dscp/DB/SQLite/Types.hs]: Okay, got it, the problem was simple. I was simply confused by the fact you used type synonyms for lists representing branches, because it felt natural to me to just inline them, and I thought that you for some reason _had_ to use them, which made no sense. Now you've just inlined the inner lists with branches and everything became clear.

**Review by flyingleafe — APPROVED** [review]
LGTM now. Please, rebase against master and make sure everything builds.

**Comment by pasqu4le** [comment]
rebased, `build` is successful and so are `test`s

File: core/src/Dscp/Resource/AppDir.hs [hunk 6]
```diff
@@ -1,29 +1,43 @@
...
+-- Note, there is a reason this contains the whole tree and not just its content
+-- see 'ConfigMaybe' for an explanation.
+type AppDirParam =
+   '[ "param" ::+
+       '[ "os" ::- '[]
+          -- ^ Dedicated folder inside OS directory for applications
+        , "specific" ::- '[ "path" ::: FilePath ]
+          -- ^ Given path
+        ]
+    ]
```
Comment (Martoon-00) [inline core/src/Dscp/Resource/AppDir.hs]: :heart:
